src/uct/base/uct_mem.c

```diff
@@ -120,17 +139,24 @@ ucs_status_t uct_mem_alloc(void *addr, size_t min_length, unsigned flags,
                 address = addr;
                 status = uct_md_mem_alloc(md, &alloc_length, &address, flags,
                                           alloc_name, &memh);
-                if (status != UCS_OK) {
+                if ((status == UCS_ERR_NO_RESOURCE) &&
+                    (flags & UCT_MD_MEM_FLAG_ON_DEVICE)) {
+                    ucs_debug("failed to allocate memory on device using md %s for %s: %s. "
+                              "trying to fallback into regular mode",
```

Comment:
need to fallback only if all MDs failed to allocate with DM flag..

Reply:
right, flags & UCT_MD_MEM_FLAG_ON_DEVICE - evaluation for revice memory requested

Reply:
but it seems like you remove UCT_MD_MEM_FLAG_ON_DEVICE from flags, as soon as one MD fails to allocate DM..
and then ALLOCATE flag is not even added..
seems like fallback logic is still not right

Reply:
this is uct_mem_alloc function... are you sure it should have ALLOC flag?

Reply:
BTW, there is no ALLOC flag in uct_md_mem_flags